### DISCUSSION TITLE
@rollup/plugin-commonjs and @rollup/plugin-babel order problems

### DISCUSSION THREAD

**Discussion by jianxingyao** [rollup/rollup#5731] [Q&A]
This is stated in the rollup documentation
![image](https://github.com/user-attachments/assets/017d1162-27fd-4def-885b-0bd98835c606)
@rollup/plugin-babel is stated as such
![image](https://github.com/user-attachments/assets/7a3f0bc2-a298-4c65-8251-3fd73bd4a0a0)

The order of the actual measurements is inconsistent, but the results are the same.
How do I use it

**Comment by tt-a1i** [comment]
The documentation might seem contradictory, but here's the key insight:

**TL;DR: Put `@rollup/plugin-commonjs` BEFORE `@rollup/plugin-babel`**

```js
import commonjs from '@rollup/plugin-commonjs';
import babel from '@rollup/plugin-babel';
import resolve from '@rollup/plugin-node-resolve';

export default {
  plugins: [
    resolve(),
    commonjs(),  // First: convert CJS to ESM
    babel({ babelHelpers: 'bundled' }),  // Then: transform with Babel
  ]
};
```

**Why this order?**

1. **node-resolve** first - resolves `import 'lodash'` to the actual file path in node_modules
2. **commonjs** second - converts CommonJS modules (`require`/`module.exports`) to ES modules
3. **babel** third - transforms the now-ESM code

The babel plugin docs say to put it "before other plugins transform your modules" - this refers to plugins that might transform your *output* (like terser/minify). The commonjs plugin transforms *input* (converting CJS dependencies to ESM), so it should run before babel.

**The confusion:**
- babel transforms your *source code* syntax (JSX, TypeScript, modern JS)
- commonjs transforms *module format* (CJS → ESM)

They operate on different concerns, and commonjs needs to run first so babel sees proper ES module syntax.

**Comment by MukundaKatta** [comment]
The order that works in practice: `commonjs` → `babel` → `nodeResolve` (or `nodeResolve` first if you need it for `.mjs` modules). The docs contradict because they're describing two different concerns: plugin-commonjs needs to run *before* babel so babel sees ES modules, but it needs `nodeResolve` earlier to actually find the cjs files. Put all three in that order and you avoid both failure modes.
